### PR TITLE
refactor: implement fake repo and inject in test using factory pattern

### PR DESCRIPTION
**Pull request by kdvuong**
- created a CustomTestRunner that set build config to testing mode.
- Fleshed out fake repos
- inject them into the app using factory.

### CODE REVIEW

File: app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java [hunk 1]
```diff
@@ -54,28 +56,13 @@ public static BookRepositoryImpl getInstance() {
...
-                                    return new Book(newBookId, isbn, title, author, description, owner, BOOK_STATUS.AVAILABLE, addImageTask.getResult());
-                                }
-                                throw new Exception("add Book failed: updating book with new image download uri failed, everything else worked");
-                            });
-                        }
-                        throw new Exception("add Book failed: upload image failed, add image task is not successful");
-                    });
-                }
-                throw new Exception("add Book failed: failed to add new book data to collection");
-            });
```
Comment (kdvuong) [inline app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java:74]: ignore this, this is from another branch.

File: app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java [hunk 2]
```diff
@@ -114,95 +102,74 @@ public static BookRepositoryImpl getInstance() {
...
-                        });
-                    } else {
-                        // just return the updated book
-                        return bookCollection.get().continueWith(noop -> newBook);
-                    }
-                }
-                throw new Exception("edit Book failed: failed to update with data");
-            });
-        }
-
```
Comment (kdvuong) [inline app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java:167]: ignore here too.

File: app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java [hunk 3]
```diff
@@ -114,95 +102,74 @@ public static BookRepositoryImpl getInstance() {
...
-        return new Book(id, isbn, title, author, description, owner, BOOK_STATUS.valueOf(status));
-    }
-
-    public Task<Uri> getImage(String imageName) {
-        StorageReference pathReference = imageStorageReference.child(imageName);
-        return pathReference.getDownloadUrl();
-    }
-
-    public Task<QuerySnapshot> getBooksOfOwnerId(String username) {
-        return bookCollection.whereEqualTo("owner", username).get();
```
Comment (kdvuong) [inline app/src/main/java/com/cmput301f20t21/bookfriends/repositories/impl/BookRepositoryImpl.java:196]: ignore

File: app/src/androidTest/java/com/cmput301f20t21/bookfriends/ui/login/CustomTestRunner.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+public final class CustomTestRunner extends Runner implements Filterable, Sortable {
+
+    private static final String TAG = "CustomTestRunner";
+
+    private final Runner delegate;
+
+    public CustomTestRunner(Class<?> klass) throws InitializationError {
+        delegate = loadRunner(klass);
+        BuildConfig.IS_TESTING.set(true);
```
Comment (kdvuong) [inline app/src/androidTest/java/com/cmput301f20t21/bookfriends/ui/login/CustomTestRunner.java:33]: I copied this class from `AndroidJUnit4.class`, I wanted to extends it and add line 33 but it didn't let me.

**Comment by kdvuong** [comment]
I haven't implemented FakeRequestRepository in this PR.

File: app/src/androidTest/java/com/cmput301f20t21/bookfriends/ui/login/CustomTestRunner.java (hunk 4 above)
Comment (c25vdw) [reply app/src/androidTest/java/com/cmput301f20t21/bookfriends/ui/login/CustomTestRunner.java:33]: question: what's the purpose of a new Test Runner class? were we not able to set the IS_TESTING to false in a normal test case?

Comment (kdvuong) [reply app/src/androidTest/java/com/cmput301f20t21/bookfriends/ui/login/CustomTestRunner.java:33]: we can but every test we will need to set it. Having a custom test runner we can abstract that, more convenient (maybe in the future we want to change the way we set IS_TESTING, this approach we be easier to modify).

**Review by c25vdw — APPROVED** [review]
lgtm. with this PR merged we need to tweak all intent tests to add data ourselves.

**Comment by kdvuong** [comment]
> lgtm. with this PR merged we need to tweak all intent tests to add data ourselves.

right now none of our intent test rely on data from the database except login i think.

**Review by zhpeng811 — APPROVED** [review]
lgtm

**Pull request merged**